### PR TITLE
CP-41972: update xcp to support ack py3

### PR DESCRIPTION
**Pull request by alexhimmel**

### CODE REVIEW

File: xcp/pci.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def __init__(self, fn):
         vendor = None
         cls = None
 
-        fh = open(fn)
+        fh = open(fn, 'r', encoding='UTF-8')
         for l in fh:
             line = l.rstrip()
```
Comment (bernhardkaindl) [inline xcp/pci.py:191]: The purpose of this change is not clear, the reason is:
> open() does not change it's return value between Python2 and Python3:
> See: https://docs.python.org/3.6/library/functions.html#open -> See "text mode (default)"

All of them for just fine and produce the same output:
```py
python2 -c 'file=open("/etc/passwd"); lines = [line.decode() for line in file]; print(lines[0])'
python3 -c 'file=open("/etc/passwd"); lines = [line.decode() for line in file]; print(lines[0])'
python3 -c 'file=open("/etc/passwd", "r", encoding="UTF-8"); lines = [line for line in file]; print(lines[0])'
```
But `encoding=` is not supported by Python2:
```py
python2.7 -c "open('/etc/passwd', 'r', encoding='UTF-8')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: 'encoding' is an invalid keyword argument for this function
```
Therefore, revert this change:
```suggestion
        fh = open(fn)
        for l in fh:
            line = l.rstrip()
```

File: xcp/pci.py [hunk 2]
```diff
@@ -260,7 +260,7 @@ def __init__(self):
                                stdout = subprocess.PIPE)
         for l in cmd.stdout:
             line = l.rstrip()
-            el = [x for x in line.replace('"', '').split() if not x.startswith('-')]
```
Comment (bernhardkaindl) [inline xcp/pci.py:263]: Thanks, this is a correct fix, because Popen() changes it's return value to byes.

Because the testsuite is not mocking this change in the return value of Popen() yet, this would break the test suite. It has to be changed like this:

tests/test_pci.py:
```diff
+        def fake_Popen():
+            """Mimic Popen().output: Python3 returns bytes, Python2 returns str"""
+            return open("tests/data/lspci-mn", "rb" if sys.version_info[0] == 3 else "r")
+
         with patch("xcp.pci.subprocess.Popen") as popen_mock, \
-             open("tests/data/lspci-mn") as fake_data:
+            fake_Popen() as fake_data:
             popen_mock.return_value.stdout.__iter__ = Mock(return_value=iter(fake_data))
             devs = PCIDevices()
```
I hope I can submit this tomorrow if you haven't by then.

Likely likewise for the Popen() for biosdevname, if it has a test.
If not, a test for it should be added.

Comment (tianxiahcp) [reply xcp/pci.py:263]: Hi @bernhardkaindl 
Thank you for helping me to review this PR, i will test with new branch.
Regards,
Tian

Comment (alexhimmel) [reply xcp/pci.py:263]: Hi @bernhardkaindl 
Thank you for helping me to review this PR, i will test with new branch "testsuite-driven-py3+Popen".
Regards,
Tian

File: xcp/pci.py (hunk 1 above)
Comment (alexhimmel) [reply xcp/pci.py:191]: Hi @bernhardkaindl 
Got error when try the nwe branch with `fh = open(fn)`
`
About to call plugin 'get_network_devices' on host 'OpaqueRef:482bcea2-6914-9d49-ecc9-6708b236fa9e' with args '{}'
Traceback (most recent call last):
  File "./ack_cli.py", line 536, in <module>
    main(config, test_run_file)
  File "./ack_cli.py", line 508, in main
    generate_test_config(session, config, test_run_file)
  File "./ack_cli.py", line 404, in generate_test_config
    ifs = network_interfaces_to_test(session, config)
  File "./ack_cli.py", line 335, in network_interfaces_to_test
    devices = utils.get_master_network_devices(session)
  File "/opt/xensource/packages/files/auto-cert-kit/utils.py", line 2427, in get_master_network_devices
    nics = call_ack_plugin(session, 'get_network_devices')
  File "/opt/xensource/packages/files/auto-cert-kit/utils.py", line 2226, in call_ack_plugin
    args)
  File "/usr/lib/python3.6/site-packages/XenAPI.py", line 260, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python3.6/site-packages/XenAPI.py", line 154, in xenapi_request
    result = _parse_result(getattr(self, methodname)(*full_params))
  File "/usr/lib/python3.6/site-packages/XenAPI.py", line 234, in _parse_result
    raise Failure(result['ErrorDescription'])
XenAPI.Failure: ['XENAPI_PLUGIN_FAILURE', 'get_network_devices', 'UnicodeDecodeError', "'ascii' codec can't decode byte 0xc2 in position 265: ordinal not in range(128)"]
`
Can fix this issue when specify the correct codec when opening the file or when calling the decode method e.g.
`fh = open(fn, 'r', encoding='UTF-8')` or  somewhere decode the `stdout` with `stdout.decode(''UTF-8)`

File: xcp/pci.py (hunk 2 above)
Comment (bernhardkaindl) [reply xcp/pci.py:263]: The fixes issues which the changes in this PR adressed are now merged to master. 

Hi @alexhimmel, you could schedule re-testing the ACK XAPI plugin with the new master branch in an upcoming sprint!

PS: The Python3 migration is not done yet, there are still many issues (or at least untested cases) which need testing and likely fixing.

For example the github action summay page is currently still showing pytype errors in xcp/accessor.py and xcp/repository.py.

If the ACK plugin uses such modules or you run into any other error, please open an issue here to report the problem, thanks!

**Comment by bernhardkaindl** [comment]
@alexhimmel tested https://github.com/xenserver/auto-cert-kit with the fixes from my merged PRs.
